src/Conversion/ONNXToKrnl/ML/CategoryMapper.cpp

```diff
@@ -0,0 +1,294 @@
+    ArrayAttr cats_int64sAttr = categoryMapperOp.cats_int64sAttr();
+    ArrayAttr cats_stringsAttr = categoryMapperOp.cats_stringsAttr();
+
+    DenseElementsAttr cats_int64s = mlir::DenseElementsAttr::get(
+        RankedTensorType::get(
+            cats_int64sAttr.size(), rewriter.getIntegerType(64)),
+        cats_int64sAttr.getValue());
+    DenseElementsAttr cats_strings = mlir::DenseElementsAttr::get(
+        RankedTensorType::get(
+            cats_stringsAttr.size(), StringType::get(rewriter.getContext())),
+        cats_stringsAttr.getValue());
+
+    IntegerAttr default_int64 = categoryMapperOp.default_int64Attr();
+    DenseElementsAttr default_string =
+        (categoryMapperOp.default_stringAttr())
+            ? mlir::DenseElementsAttr::get(
+                  RankedTensorType::get(
+                      {}, StringType::get(rewriter.getContext())),
+                  categoryMapperOp.default_stringAttr().getValue())
+            : nullptr;
+
+    // Basic information.
+    auto memRefType = convertToMemRefType(*op->result_type_begin());
+    int64_t rank = memRefType.getShape().size();
+    ShapedType inputType = X.getType().cast<ShapedType>();
+    Type elementType = inputType.getElementType();
+
+    // Insert an allocation and deallocation for the result of this operation.
+    Value alloc = insertAllocAndDeallocSimple(
+        rewriter, op, memRefType, loc, shapeHelper.dimsForOutput(0));
+
+    KrnlBuilder createKrnl(rewriter, op->getLoc());
+    MathBuilder createMath(createKrnl);
+
+    // Generate a perfect hash table. The hash table will be used to lookup the
+    // index of the input values.
+    PerfectHashTable perfectHashTable =
+        createPerfectHashTable(cats_int64s, cats_strings, cats_int64sAttr,
+            cats_stringsAttr, elementType, createKrnl);
+
+    // Create loop invariant values.
+    Value constantForCatsInt64s = createKrnl.constant(
+        convertToMemRefType(cats_int64s.getType()), "cats_int64s", cats_int64s);
+
+    llvm::errs() << "constantForCatsInt64s: " << constantForCatsInt64s << "\n";
+    Value constantForCatsStrings =
+        createKrnl.constant(convertToMemRefType(cats_strings.getType()),
+            "cats_strings", cats_strings);
+
+    Value defaultInt64 = (default_int64)
+                             ? createMath.constant(rewriter.getIntegerType(64),
+                                   default_int64.getSInt())
+                             : nullptr;
+    Value defaultString =
+        (default_string)
+            ? createKrnl.constant(
+                  MemRefType::get({}, StringType::get(rewriter.getContext())),
+                  "default_string", default_string)
+            : nullptr;
+
+    // Lookup the index in the perfect hash table corresponding to
+    // each input value.
+    BuildKrnlLoop inputLoops(rewriter, loc, rank);
+    inputLoops.createDefineAndIterateOp(X);
+    rewriter.setInsertionPointToStart(inputLoops.getIterateBlock());
```

Comment:
This style of building `krnl.iterate` is quite old. Could we change to the new style that uses lambda function: https://github.com/onnx/onnx-mlir/blob/master/docs/LoweringCode.md?

Reply:
Done.